NullPointerException on updating hierarchical watches

```
Caused by: org.gradle.api.GradleException: Build cache entry 6ce9752b1f625341974870b5e810c309 from local build cache is invalid
	at org.gradle.caching.internal.controller.DefaultBuildCacheController.load(DefaultBuildCacheController.java:103)
	at org.gradle.internal.execution.steps.CacheStep.lambda$executeWithCache$0(CacheStep.java:78)
	at org.gradle.internal.Try.ofFailable(Try.java:39)
	at org.gradle.internal.execution.steps.CacheStep.executeWithCache(CacheStep.java:77)
	... 57 more
Caused by: java.lang.NullPointerException
	at org.gradle.internal.watch.vfs.impl.DelegatingDiffCapturingUpdateFunctionDecorator.lambda$decorate$2(DelegatingDiffCapturingUpdateFunctionDecorator.java:56)
	at org.gradle.internal.snapshot.AtomicSnapshotHierarchyReference.update(AtomicSnapshotHierarchyReference.java:38)
	at org.gradle.internal.vfs.impl.DefaultVirtualFileSystem.updateRoot(DefaultVirtualFileSystem.java:155)
	at org.gradle.internal.vfs.impl.DefaultVirtualFileSystem.updateWithKnownSnapshot(DefaultVirtualFileSystem.java:193)
	at org.gradle.internal.watch.vfs.impl.AbstractDelegatingVirtualFileSystem.updateWithKnownSnapshot(AbstractDelegatingVirtualFileSystem.java:63)
	at org.gradle.internal.watch.vfs.impl.WatchingVirtualFileSystem.updateWithKnownSnapshot(WatchingVirtualFileSystem.java:51)
	at org.gradle.internal.vfs.RoutingVirtualFileSystem.updateWithKnownSnapshot(RoutingVirtualFileSystem.java:95)
	at org.gradle.caching.internal.controller.impl.DefaultBuildCacheCommandFactory$LoadCommand.lambda$snapshotUnpackedData$2(DefaultBuildCacheCommandFactory.java:140)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution.visitOutputTrees(ExecuteActionsTaskExecuter.java:352)
	at org.gradle.caching.internal.controller.impl.DefaultBuildCacheCommandFactory$LoadCommand.snapshotUnpackedData(DefaultBuildCacheCommandFactory.java:122)
	at org.gradle.caching.internal.controller.impl.DefaultBuildCacheCommandFactory$LoadCommand.load(DefaultBuildCacheCommandFactory.java:95)
	at org.gradle.caching.internal.controller.DefaultBuildCacheController$Unpack$1.run(DefaultBuildCacheController.java:150)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:395)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:387)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:157)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:242)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:150)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:84)
	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
	at org.gradle.caching.internal.controller.DefaultBuildCacheController$Unpack.execute(DefaultBuildCacheController.java:146)
	at org.gradle.caching.internal.controller.DefaultBuildCacheController$Unpack.execute(DefaultBuildCacheController.java:135)
	at org.gradle.caching.local.internal.DirectoryBuildCacheService.loadInsideLock(DirectoryBuildCacheService.java:117)
	at org.gradle.caching.local.internal.DirectoryBuildCacheService.access$200(DirectoryBuildCacheService.java:42)
	at org.gradle.caching.local.internal.DirectoryBuildCacheService$1.run(DirectoryBuildCacheService.java:99)
	at org.gradle.internal.Factories$1.create(Factories.java:26)
	at org.gradle.cache.internal.LockOnDemandCrossProcessCacheAccess.withFileLock(LockOnDemandCrossProcessCacheAccess.java:90)
	at org.gradle.cache.internal.DefaultCacheAccess.withFileLock(DefaultCacheAccess.java:196)
	at org.gradle.cache.internal.DefaultPersistentDirectoryStore.withFileLock(DefaultPersistentDirectoryStore.java:182)
	at org.gradle.cache.internal.DefaultCacheFactory$ReferenceTrackingCache.withFileLock(DefaultCacheFactory.java:206)
	at org.gradle.caching.local.internal.DirectoryBuildCacheService.loadLocally(DirectoryBuildCacheService.java:94)
	at org.gradle.caching.internal.controller.service.DefaultLocalBuildCacheServiceHandle.load(DefaultLocalBuildCacheServiceHandle.java:49)
	at org.gradle.caching.internal.controller.DefaultBuildCacheController.load(DefaultBuildCacheController.java:101)
	... 60 more
```

See 
https://builds.gradle.org/buildConfiguration/Gradle_Check_VfsRetention_30_bucket18/34841091?buildTab=overview

---
cc: @gradle/build-cache

--- a/subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/DelegatingDiffCapturingUpdateFunctionDecorator.java
+++ b/subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/DelegatingDiffCapturingUpdateFunctionDecorator.java
@@ -26,39 +26,50 @@ import java.util.function.Predicate;
 public class DelegatingDiffCapturingUpdateFunctionDecorator implements SnapshotHierarchy.DiffCapturingUpdateFunctionDecorator {
 
     private final Predicate<String> watchFilter;
-    private SnapshotHierarchy.SnapshotDiffListener snapshotDiffListener;
-    private ErrorHandler errorHandler;
+    private ErrorHandlingDiffPublisher errorHandlingDiffPublisher;
 
     public DelegatingDiffCapturingUpdateFunctionDecorator(Predicate<String> watchFilter) {
         this.watchFilter = watchFilter;
     }
 
     public void setSnapshotDiffListener(SnapshotHierarchy.SnapshotDiffListener snapshotDiffListener, ErrorHandler errorHandler) {
-        this.snapshotDiffListener = snapshotDiffListener;
-        this.errorHandler = errorHandler;
+        this.errorHandlingDiffPublisher = new ErrorHandlingDiffPublisher(snapshotDiffListener, errorHandler);
     }
 
     public void stopListening() {
-        this.snapshotDiffListener = null;
-        this.errorHandler = null;
+        errorHandlingDiffPublisher = null;
     }
 
     @Override
     public AtomicSnapshotHierarchyReference.UpdateFunction decorate(SnapshotHierarchy.DiffCapturingUpdateFunction updateFunction) {
-        SnapshotHierarchy.SnapshotDiffListener currentListener = snapshotDiffListener;
-        if (currentListener == null) {
+        ErrorHandlingDiffPublisher currentErrorHandlingDiffPublisher = errorHandlingDiffPublisher;
+        if (currentErrorHandlingDiffPublisher == null) {
             return root -> updateFunction.update(root, SnapshotHierarchy.NodeDiffListener.NOOP);
         }
 
         SnapshotCollectingDiffListener diffListener = new SnapshotCollectingDiffListener(watchFilter);
         return root -> {
             SnapshotHierarchy newRoot = updateFunction.update(root, diffListener);
-            return errorHandler.handleErrors(newRoot, () -> diffListener.publishSnapshotDiff(currentListener));
+            return currentErrorHandlingDiffPublisher.publishSnapshotDiff(diffListener, newRoot);
         };
     }
 
     public interface ErrorHandler {
         @CheckReturnValue
         SnapshotHierarchy handleErrors(SnapshotHierarchy currentRoot, Runnable runnable);
+    }
+
+    private static class ErrorHandlingDiffPublisher {
+        private final SnapshotHierarchy.SnapshotDiffListener diffListener;
+        private final ErrorHandler errorHandler;
+
+        public ErrorHandlingDiffPublisher(SnapshotHierarchy.SnapshotDiffListener diffListener, ErrorHandler errorHandler) {
+            this.diffListener = diffListener;
+            this.errorHandler = errorHandler;
+        }
+
+        public SnapshotHierarchy publishSnapshotDiff(SnapshotCollectingDiffListener collectedDiff, SnapshotHierarchy newRoot) {
+            return errorHandler.handleErrors(newRoot, () -> collectedDiff.publishSnapshotDiff(diffListener));
+        }
     }
 }

